Gleam-from-Elm cheatsheet

michaeljones opened the pull request:
Starting a PR in case there is any early thoughts or feedback. No rush. I'll try to work through the rest this week.

---

Starting from the Elixir cheatsheet at Louis' recommendation. I've added
an overview to give a broad comparison before proceeding to edit the
document line by line to shape to Elm instead.

It is not completed. I have removed the unchanged parts for the moment
though the entries in the table of contents still correspond to them.

michaeljones commented:
Just pushed another commit that sees the rest of the file converted.

I guess it is less than ideal as it is still goes feature by feature but through Elixir features and then explains the relevance to Elm & Gleam (without mentioned Elixir but that is the skeleton I've built off.) It presumably makes sense to restructure based on Elm features and copy the sections in where desirable. I'll continue to work on it.

michaeljones commented:
Reorganised with a table of contents which is more Elm targeted. Plenty of TODOs though so much more work to be done. Looks more convincingly like an Elm dev's guide now though.

lpil commented on cheatsheets/gleam-for-elm-users.md:
```diff
@@ -0,0 +1,699 @@
+## Overview
+
+Elm and Gleam have similar goals of providing a robust and sound type system with an approachable
+set of features. Gleam started off with a similar syntax to Elm but as moved to use one that is more
+familiar to many developers.
```
```suggestion
set of features.
```

An interesting bit of history, but not something I think belongs here. I wouldn't want any newcomer Elm syntax fans to feel let down by this change.

lpil commented on cheatsheets/gleam-for-elm-users.md:
````diff
@@ -0,0 +1,699 @@
+const answer: Int = 42
+```
+
+There are no multiline commends in Gleam.
````
```suggestion
There are no multiline comments in Gleam.
```

lpil commented on cheatsheets/gleam-for-elm-users.md:
````diff
@@ -0,0 +1,699 @@
+// Hello, Joe!
+```
+
+Comments starting with `///` are used to document the following statement. Comments starting with `////` are used to document the current module.
````
```suggestion
Comments starting with `///` are used to document the following function, constant, or type definition. Comments starting with `////` are used to document the current module.
```

There was some confusion around the word "statement" in discord the other day. I am hoping this language is clearer.

lpil commented on cheatsheets/gleam-for-elm-users.md:
```diff
@@ -0,0 +1,699 @@
+#### Gleam
+
+Gleam has the `let` keyword before its variable names. You can re-assign variables and you can
+shadow variables from other scopes.
```
```suggestion
shadow variables from other scopes. This does not mutate the previously assigned value.
```

lpil commented on cheatsheets/gleam-for-elm-users.md:
```diff
@@ -0,0 +1,699 @@
+sum : number -> number -> number
+sum x y = x + y
+
+@spec mul(number, number) :: boolean # no Elixir compile error
```
```suggestion
```

lpil commented on cheatsheets/gleam-for-elm-users.md:
````diff
@@ -0,0 +1,699 @@
+}
+```
+
+### Function heads
+
+Neither Elm nor Gleam support multiple function heads like Erlang, Elixir or Haskell. Both languages
+would expect a case-statement to be used to handle variations in data being supplied to the
+function.
+
+### Function overloading
+
+Neither Elm nor Gleam support function overloading.
+
+### Referencing functions
+
+Both Elm and Gleam have a single namespace for value and functions within a module, so there
+is no need for a special syntax to assign a module function to a variable.
+
+### Calling anonymous functions
+
+Neither Elm nor Gleam require special syntax for calling anonymous functions.
+
````
Maybe we can remove this bit? Not sure Elm programmers would be expecting these features. What do you think?

```suggestion
```

michaeljones replied:
Yeah, makes total sense. This a hang over from the original Elixir layout and it seemed odd to delete valid information but I can see it is potentially confusing if someone knows little about these things.

lpil commented on cheatsheets/gleam-for-elm-users.md:
```diff
@@ -0,0 +1,699 @@
+
+#### Gleam
+
+There is no limit to the number of entries in Gleam tuples.
```
```suggestion
There is no limit to the number of entries in Gleam tuples, but records are still recommended as giving names to fields adds clarity.
```

lpil commented on cheatsheets/gleam-for-elm-users.md:
````diff
@@ -0,0 +1,699 @@
+
+let person = Person(name: "Jake", age: 35)
+let name = person.name
+```
````
Not sure what is happening here, GitHub's syntax highlighting seems to have broken?

lpil reviewed (commented):
This is fantastic work, thank you. Your writing is very clear. I've left some small notes but otherwise I'd love to merge this in when you're happy.

michaeljones commented:
I've added your changes, completed the outstanding TODO sections and removed line breaks from paragraphs for consistency and to reduce diff noise on future edits. Though personally, I've not figured out how to get vim to wrap the lines at a reason width so I find this less readable :) 

I'm not the best at proof-reading my own work so edits are welcome.

I thought we might add a section on `todo` at some point as `Debug.todo` and `todo` exist as parallels. Not aiming to include that here myself.

I have attempted to link to the language tour in one place (for opaque types.) I'm not sure if the syntax of the link will work properly. 

Otherwise I'm happy with it to be merged if you're happy.

lpil commented:
> I thought we might add a section on todo at some point as Debug.todo and todo exist as parallels. Not aiming to include that here myself.

Ah yes! That would be a good addition. Thank you
